pkg/pubsub/pubsub_init.go

```diff
@@ -0,0 +1,57 @@
+package pubsub
+
+import (
+	"context"
+
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+var log = logf.Log.WithName("pubsub")
+
+var _ manager.Runnable = &runner{}
+
+type runner struct {
+	mgr manager.Manager
+}
+
+func AddToManager(m manager.Manager) error {
+	mr := new(m)
+	return m.Add(mr)
+}
+
+func new(mgr manager.Manager) *runner {
+	mr := &runner{
+		mgr: mgr,
+	}
+	return mr
+}
+
+// Start implements the Runnable interface.
+func (r *runner) Start(ctx context.Context) error {
+	errCh := make(chan error)
+	tools := Tools()
+	if len(tools) == 0 {
+		log.Info("No pub sub tool is enabled")
+		return nil
+	}
+	log.Info("initializing pub subs")
+	client := r.mgr.GetClient()
+	for i := range tools {
+		pubsub := tools[i]
+		go func() {
+			if err := pubsub.NewClient(ctx, client); err != nil {
```

Comment:
What happens if the auth credentials/config are not available at startup time?

Reply:
Added a default config option as well, in case config is missing at start up time

Reply:
My point is this:

If we are creating config maps/secrets/CRDs to configure these pubsub objects, then they may very well be added after Gatekeeper has started. Since this bootstrapping is only done on Gatekeeper startup, this doesn't work. Basically, we need a controller to be watching these configs and updating pubsub as they are created/modified (unless we are okay with restarting Gatekeeper for every config update).

Reply:
Added a config controller and a reconciler to initiate and update clients when respective resources are created